### PR TITLE
Refactor date forms

### PR DESCRIPTION
**Pull request by jonnyshaw89**
### What is the context of this PR?
Removed duplication from validators
Dates functional test was commented out, so this had been uncommented and fixed

### How to review 
Do dates still work as expected.
Is the refactor of the date form good enough?

### CODE REVIEW

File: app/forms/date_form.py [hunk 1]
```diff
@@ -16,191 +16,125 @@
...
-        super(DateField, self).__init__(form_class, **kwargs)
-
-    def process(self, formdata, data=None):
-        if data is not None:
-            substrings = data.split('-')
-            data = {'year': substrings[0], 'month': substrings[1], 'day': substrings[2]}
-
-        super().process(formdata, data)
-
+class DateFormTypes(Enum):
```
Comment (bitdivision) [inline app/forms/date_form.py]: :+1:

Though, I think this should be called `DateFormType`. It represents a single type.

File: app/forms/date_form.py [hunk 2]
```diff
@@ -16,191 +16,125 @@
...
+        metadata,
+        answer,
+        error_messages,
+        **kwargs,
+    ):
+        form_class = get_form(
+            date_form_type, answer, answer_store, metadata, error_messages
         )
-        super(YearField, self).__init__(form_class, **kwargs)
+        super(DateField, self).__init__(form_class, **kwargs)
```
Comment (bitdivision) [inline app/forms/date_form.py]: You should be able to use `super().__init__(form_class, **kwargs)`. Using the name of the parent class is not needed in python 3.

File: app/forms/date_form.py [hunk 3]
```diff
@@ -16,191 +16,125 @@
...
-    """
-    Returns a month year form metaclass with appropriate validators. Used in both date and
-    date range form creation.
+class DateForm(Form):
+    @CachedProperty
+    def data(self):
+        data = super().data
+        try:
+            if all(k in data for k in ('day', 'month', 'year')):
+                return '{year:04d}-{month:02d}-{day:02d}'.format(
```
Comment (bitdivision) [inline app/forms/date_form.py:89]: Might be a bit cleaner:
                `return f'{data['year']:04d}-{data['month']:02d}-{data['day']:02d}'`

File: app/forms/date_form.py [hunk 4]
```diff
@@ -16,191 +16,125 @@
...
+                    month=int(data['month']),
+                    day=int(data['day']),
+                )
 
-    :param answer: The answer on which to base this form
-    :param error_messages: The messages to use upon this form during validation
-    :return: The generated MonthYearDateForm metaclass
-    """
+            if all(k in data for k in ('month', 'year')):
+                return '{year:04d}-{month:02d}'.format(
```
Comment (bitdivision) [inline app/forms/date_form.py:96]: Same again: `return f'{data['year']:04d}-{data['month']:02d}'`

File: app/forms/date_form.py [hunk 5]
```diff
@@ -16,191 +16,125 @@
...
+                    year=int(data['year']), month=int(data['month'])
+                )
 
-    class MonthYearDateForm(Form):
-        year = StringField()
-        month = StringField()
+            if 'year' in data and data['year']:
+                return '{year:04d}'.format(year=int(data['year']))
```
Comment (bitdivision) [inline app/forms/date_form.py:101]: `return f'{data['year']:04d}'`

File: app/forms/date_form.py [hunk 6]
```diff
@@ -16,191 +16,125 @@
...
-        def data(self):
-            data = super().data
-            if not data['year']:
-                return None
+    class CustomDateForm(DateForm):
+        pass
 
-            return '{:04d}'.format(int(data['year']))
+    if form_type in DateFormTypes:
+        # Validation os only ever added to the 1 field that shows in all 3 variants
```
Comment (bitdivision) [inline app/forms/date_form.py]:
```suggestion
        # Validation is only ever added to the 1 field that shows in all 3 variants
```

**Review by bitdivision — CHANGES_REQUESTED** [review]
This is looking much much better.

And thanks for fixing my tests :)

File: app/forms/date_form.py (hunk 3 above)
Comment (jonnyshaw89) [reply app/forms/date_form.py:89]: would it not have to be 
`return f'{int(data['year']):04d}-{int(data['month']):02d}-{int(data['day']):02d}'`

Comment (bitdivision) [reply app/forms/date_form.py:89]: Yeah, you're right. I guess we could cast those in one place, but doesn't make much difference overall.

**Review by bitdivision — APPROVED** [review]

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**